app/org/sagebionetworks/bridge/upload/UploadValidationTask.java

```
@@ -111,6 +122,61 @@ public void run() {
         }
 
         // TODO: if validation fails, wipe the files from S3
+
+        // dedupe logic over here:
+        try {
+            dedupeHelper(context.getRecordId());
+        } catch (ClassNotFoundException e) {
+            logClassNotFoundException(e);
+        }
+
+    }
+
+    // helper method to log exception
+    void logClassNotFoundException(ClassNotFoundException e) {
+        logger.error("Record not found in ddb", e);
+    }
+
+    /**
+     * helper method: query healthdatarecord by healthcode, schema and createdOn from the upload just finished above,
+     * if there are more than 1 such records in ddb, there are duplicates. log that information but not stop or delete anything right now
+     * @param healthRecordId
+     */
+    private void dedupeHelper(String healthRecordId) throws ClassNotFoundException {
+        // get necessary information for query first
+        HealthDataRecord record = healthDataService.getRecordById(healthRecordId);
+
+        // if there is no record in ddb,
+        if (record == null) {
+            throw new ClassNotFoundException("Record not found in ddb");
+        }
+
+        Long createdOn = record.getCreatedOn();
+        String healthCode = record.getHealthCode();
+        String schemaId = record.getSchemaId();
+
+        List<HealthDataRecord> retList = healthDataService.getRecordsByHealthcodeCreatedOnSchemaId(healthCode, createdOn, schemaId);
+
+        if (retList.size() > 1) {
+            logger.info(String.format("Duplicate health data records for record id: %s, health code: %s, created on:  %s, schema id: %s, duplicate size: %s",
+                    healthRecordId, healthCode, createdOn, schemaId, retList.size()));
+
+            logDuplicateUploadRecords(retList);
+        } else if (retList.size() == 0) {
+            // should not be reached since getRecordById() already test if there is a record exists
+            throw new ClassNotFoundException("Record not found in ddb");
```

Do we need this else if check here? Is ClassNotFoundException the right exception to throw here? Is there any particular reason we're throwing an exception instead of logging?